src/mongocxx/logger.hpp

```diff
@@ -0,0 +1,66 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <memory>
+
+#include <bsoncxx/stdx/string_view.hpp>
+
+#include <mongocxx/stdx.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+///
+/// The log level of a message passed to a @logger.
+///
+enum class log_level {
+    k_error,
+    k_critical,
+    k_warning,
+    k_message,
+    k_info,
+    k_debug,
+    k_trace,
+};
+
+///
+/// The interface that all user-defined loggers must implement.
+///
+class MONGOCXX_API logger {
+   public:
+    virtual ~logger() = default;
```

Comment:
I feel like we want the =default in the .cpp file for ABI reasons? For the ctor as well?